File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
    let prev = null;
    let count = 0;
    let max = 0;
    let result = [];

    const dfs = function(node) {
        if(!node) return;

        dfs(node.left);

        if(prev !== null && prev.val === node.val) {
            count++;
        } else {
            count = 1;
        }

        if(max < count) {
            result = [node.val];
            max = count;
        } else if(max === count) {
            result.push(node.val);
        }

        prev = node;

        dfs(node.right);
    }

    dfs(root);

    return result;
};

File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        prev = None
        count = 0
        max = 0
        result = []

        def dfs(node):
            nonlocal prev, count, max, result

            if node is None:
                return

            dfs(node.left)

            if prev is not None and prev.val == node.val:
                count += 1
            else:
                count = 1

            if max < count:
                result = [node.val]
                max = count
            elif max == count:
                result.append(node.val)

            prev = node

            dfs(node.right)

        dfs(root)

        return result
